src/org/omegat/util/gui/FontFallbackManager.java

```diff
@@ -52,15 +55,97 @@ private FontFallbackManager() {
+            Arrays.asList("Apple Color Emoji", "Noto Color Emoji")
+    );
+    private static final Font FONT_UNAVAILABLE = new Font("", Font.PLAIN, 0);
 
     private static final Logger LOGGER = Logger.getLogger(FontFallbackManager.class.getName());
 
     private static final Font[] RECENT_FONTS = new Font[8];
     private static int lastFontIndex = 0;
     private static final Map<Integer, Font> CACHE = new ConcurrentHashMap<>();
 
+    private static final FontRenderContext DEFAULT_CONTEXT = new FontRenderContext(null, false, false);
+
+    private FontFallbackManager() {
+    }
+
+    /**
+     * Detect specified codePoints can display with specified font.
+     */
+    public static int canDisplayUpTo(Font font, String str) {
+        int len = str.length();
+        for (int i = 0; i < len; i++) {
+            char c = str.charAt(i);
+            if (canDisplay(font, c)) {
+                continue;
+            }
+            if (!Character.isHighSurrogate(c)) {
+                return i;
+            }
+            if (!canDisplay(font, str.codePointAt(i))) {
+                return i;
+            }
+            i++;
+        }
+        return -1;
+    }
+
+    public static int canDisplayUpTo(Font font, char[] text, int start, int limit) {
+        for (int i = start; i < limit; i++) {
+            char c = text[i];
+            if (canDisplay(font, c)) {
+                continue;
+            }
+            if (!Character.isHighSurrogate(c)) {
+                return i;
+            }
+            if (!canDisplay(font, Character.codePointAt(text, i, limit))) {
+                return i;
+            }
+            i++;
+        }
+        return -1;
+    }
+
+    public static int canDisplayUpTo(Font font, CharacterIterator iter, int start, int limit) {
+        char c = iter.setIndex(start);
+        for (int i = start; i < limit; i++, c = iter.next()) {
+            if (canDisplay(font, c)) {
+                continue;
+            }
+            if (!Character.isHighSurrogate(c)) {
+                return i;
+            }
+            char c2 = iter.next();
+            if (!Character.isLowSurrogate(c2)) {
+                return i;
+            }
+            if (!canDisplay(font, Character.toCodePoint(c, c2))) {
+                return i;
+            }
+            i++;
+        }
+        return -1;
+    }
+
+    /**
+     * Detect specified codePoint can display with specified font.
+     * @param font check against.
+     * @param codePoint character to display.
+     * @return true when specifed character can display on font, otherwise false.
+     */
+    public static boolean canDisplay(Font font, final int codePoint) {
+        if (!Character.isValidCodePoint(codePoint)) return false;
+        if (Platform.isMacOSX()) {
+            int glyphCode = font.createGlyphVector(DEFAULT_CONTEXT, new String(new int[]{codePoint}, 0, 1)).getGlyphCode(0);
+            return (0 < glyphCode && glyphCode <= 0x00ffffff);
```

Comment:
Instead of instantiating a string, you can use the method that takes an `int[]`.

Minor nitpicks:
- It's OmegaT style to always use brackets even for single-line bodies of `if` statements
- Unnecessary parens
```suggestion
        if (!Character.isValidCodePoint(codePoint)){
            return false;
        }
        if (Platform.isMacOSX()) {
            int glyphCode = font.createGlyphVector(DEFAULT_CONTEXT, Character.toChars(codePoint), 0, 1)).getGlyphCode(0);
            return 0 < glyphCode && glyphCode <= 0x00ffffff;
```

Reply:
merge the suggested.